internal/batch/batch_processor.go

```diff
@@ -568,11 +601,54 @@ func (bp *batchProcessor) dispatchBatch(payload *DispatchPayload) error {
 	// Call the dispatcher to do the heavy lifting - will only exit if we're closed
 	return operations.RunWithOperationContext(bp.ctx, func(ctx context.Context) error {
 		return bp.retry.Do(ctx, "batch dispatch", func(attempt int) (retry bool, err error) {
-			return true, bp.conf.dispatch(ctx, payload)
+			err = bp.conf.dispatch(ctx, payload)
+			if err != nil {
+				if bp.isCancelled() {
+					err = nil
+					log.L(ctx).Warnf("Batch %s was cancelled - replacing with gap fill", payload.Batch.ID)
+					for _, msg := range payload.Messages {
+						if err == nil {
+							err = bp.writeGapFill(ctx, msg)
+						}
+					}
+					payload.State = core.MessageStateCancelled
+				}
+			} else {
+				if core.IsPinned(bp.conf.txType) {
+					payload.State = core.MessageStateSent
+				} else {
+					payload.State = core.MessageStateConfirmed
+				}
+			}
+			return true, err
 		})
 	})
 }
 
+func (bp *batchProcessor) writeGapFill(ctx context.Context, msg *core.Message) error {
+	// Gap fill is only needed for private messages
+	if msg.Header.Type != core.MessageTypePrivate {
+		return nil
+	}
+	gapFill := &core.MessageInOut{
+		Message: core.Message{
+			Header:         msg.Header,
+			LocalNamespace: bp.bm.namespace,
+			State:          core.MessageStateReady,
+			Pins:           msg.Pins, // reuse any assigned pins to fill the nonce gap
+		},
+	}
+	gapFill.Header.ID = fftypes.NewUUID()
+	gapFill.Header.CID = msg.Header.ID
+	gapFill.Header.Tag = core.SystemTagGapFill
+	gapFill.Header.TxType = core.TransactionTypeBatchPin
+	err := gapFill.Seal(ctx)
+	if err == nil {
+		err = bp.data.WriteNewMessage(ctx, &data.NewMessage{Message: gapFill})
```

Comment:
One question that comes to my mind, which I wonder if you tested - what if there are other messages queued with pins after the gap-fill already?

I think the batch processor will move onto those messages and batch+send them with next-pins _after_ the gap fill, before the gap-fill arrives. So on the receiving side(s) they will receive the gap-fill out of order.

I think this just works itself out, and is all ok - but needs testing.

Reply:
I went to test this, and it does not sort itself out automatically. You can manually sort it out with another "rewind" call to the first batch that was blocked on the gap fill.

I'll try to figure out if this rewind can happen automatically. The problem is that rewinds are only effective if you rewind to an _undispatched_ pin (they are ignored if you request a rewind to a pin that has been dispatched). The gap filled pin gets marked _dispatched_ as part of this new logic - so we can't rewind to that one. We need to know if there's another pin queued up after it that is now blocked and requires a rewind...

Reply:
> that messages are confirmed on chain in the order nonces were assigned by the sender.

Yes, this is both a design assumption, and a requirement currently for a "well behaved" actor in a privacy group.

Reply:
I confirmed with @peterbroadhurst that the following is expected to be a fundamental obligation of the transaction manager: transactions will _always_ be confirmed on the chain in the order they were submitted _as long as they come from a single signing key_. This is critical for the above "well behaved" privacy group behavior to be reliable.

Therefore, all messages from a single key to a single privacy group need to be on the same thread, and therefore we have a problem in the current design where we have 2 batch processors assigning nonces from the same group (one for regular private messages, another for custom contract private messages). I think I need to combine these processors into one - can't remember the original reason I split them... likely I thought it was actually solving a problem at the time. I don't see any reason they can't be a single processor though.

Reply:
Updated so that "batch_pin" and "contract_invoke_pin" messages share a batch processor. Also altered so that the "gap fill" batch is sent out immediately during the dispatch phase, instead of trying to queue it for the next assembly loop (which had a lot of problems with reordering).